### ISSUE TITLE
Test MySQL Support

### ISSUE THREAD

**Issue by darkfrog26** [outr/scalarelational#11]
Test MySQL support.

**Comment by qbert65536** [comment]
Is this checked in ?  I'd like to test it.

**Comment by darkfrog26** [comment]
Yes, you can just extend from SQLDatastore and add the MySQL data source. If you would like to create another sub-project to test it and submit a pull request that would be awesome.

**Comment by qbert65536** [comment]
By subproject, do you mean a project similar to h2 but for mysql ?

On Tue, Jul 7, 2015 at 12:51 PM, Matt Hicks <notifications@github.com>
wrote:

> Yes, you can just extend from SQLDatastore and add the MySQL data source.
> If you would like to create another sub-project to test it and submit a
> pull request that would be awesome.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/outr/scalarelational/issues/11#issuecomment-119282067>
> .
>



-- 
Thanks!
Charlie
512-698-2009

**Comment by darkfrog26** [comment]
Yes, exactly.  I know MySQL used to have an embedded mode so you could write tests without depending on a physical install of MySQL on the machine, but otherwise it might be a bit more difficult to properly test.

**Comment by darkfrog26** [comment]
I suppose you could probably run H2 in MySQL mode (http://www.h2database.com/html/features.html#compatibility), although that would be very strange and not a flawless way of testing MySQL.

**Comment by qbert65536** [comment]
Ok I will go for adding mysql support.  Looking at the code I can see your
Scala skills are way beyond mine, if I wanted to better understand the DSL
stuff do you have any recommendations ?

On Tue, Jul 7, 2015 at 1:55 PM, Matt Hicks <notifications@github.com> wrote:

> I suppose you could probably run H2 in MySQL mode (
> http://www.h2database.com/html/features.html#compatibility), although
> that would be very strange and not a flawless way of testing MySQL.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/outr/scalarelational/issues/11#issuecomment-119302785>
> .
>



-- 
Thanks!
Charlie
512-698-2009

**Comment by darkfrog26** [comment]
The DSL is pretty straight-forward I think. Take a look at the DSLSupport trait and hopefully that will help it make more sense. Officially supporting MySQL will be a great way for ScalaRelational to be more appreciated by the community.  If you run into any problems with this please me know and I'll do my best to support getting this done.

**Comment by darkfrog26** [comment]
@qbert65536, how's it coming along?

**Comment by qbert65536** [comment]
I'll try to get more on it today, but the mysql jdbc driver doesn't like
multiple statements seperated with ';' in one execution ( I'm guessing
that's the problem - I'll paste the error shortly ).  This coming from the
create method in Datastore ( and the ddl call ).

I'll be honest I'm pretty swamped and can't dedicate a lot of time to it,
but I might be changing anorm to use scalarelational soon so I'll have to
do it at that point.

On Mon, Jul 13, 2015 at 12:13 PM, Matt Hicks <notifications@github.com>
wrote:

> @qbert65536 <https://github.com/qbert65536>, how's it coming along?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/outr/scalarelational/issues/11#issuecomment-120996128>
> .
>



-- 
Thanks!
Charlie
512-698-2009

**Comment by darkfrog26** [comment]
This shouldn't be an issue anymore. I've modified the DDL generation to separate to each statement.

**Comment by qbert65536** [comment]
Ahh, ok I'll pull and test it today.

On Mon, Jul 13, 2015 at 2:31 PM, Matt Hicks <notifications@github.com>
wrote:

> This shouldn't be an issue anymore. I've modified the DDL generation to
> separate to each statement.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/outr/scalarelational/issues/11#issuecomment-121032007>
> .
>



-- 
Thanks!
Charlie
512-698-2009

**Comment by darkfrog26** [comment]
FYI, I just added support to log the SQL before it is executed. You can simply mix-in SQLLogging into Table or Datastore and set sqlLogLevel := Level.Info to make it log all SQL as info logs.

**Comment by qbert65536** [comment]
Awesome!

On Tue, Jul 14, 2015 at 9:55 AM, Matt Hicks <notifications@github.com>
wrote:

> FYI, I just added support to log the SQL before it is executed. You can
> simply mix-in SQLLogging into Table or Datastore and set sqlLogLevel :=
> Level.Info to make it log all SQL as info logs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/outr/scalarelational/issues/11#issuecomment-121269722>
> .
>



-- 
Thanks!
Charlie
512-698-2009

**Comment by qbert65536** [comment]
How do I change default Varchar length for mysql only ?  Are you on freenode by chance ?

**Comment by darkfrog26** [comment]
Override the values on Datastore. I assume you have a MySQLDatastore that extends SQLDatastore, right?

**Comment by darkfrog26** [comment]
This can be marked as complete once TableSpec is able to successfully run.

**Comment by darkfrog26** [comment]
Should we perhaps rename this to MariaDB support instead?

**Comment by darkfrog26** [comment]
Any progress on this?

**Comment by qbert65536** [comment]
No, but we need to change the column create statement on mysql to test for
a UNIQUE index and if so, create the varchar column of under 767 bytes.

I would love to tackle it but I'm swamped all week, sorry I can't put
anymore time towards it.

On Mon, Jul 27, 2015 at 11:32 AM, Matt Hicks <notifications@github.com>
wrote:

> Any progress on this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/outr/scalarelational/issues/11#issuecomment-125265574>
> .
>



-- 
Thanks!
Charlie
512-698-2009

**Comment by lummax** [comment]
I did some stuff in [myslq-abstract-table-spec-fixes](https://github.com/lummax/scalarelational/tree/bugfix/myslq-abstract-table-spec-fixes).  Two tests are still failing, but the rest seems to work. See https://travis-ci.org/lummax/scalarelational/builds/74810356

There were some problems, all but the last were fixed.

- Some tests depend on H2 implementation details. One of them got fixed in https://github.com/outr/scalarelational/commit/09c9dc5b9882679d97691e9aca3bb228c12a4e00 (though I'm not certain, that this is the best solution). The other was fixed by introducing a `ORDER BY` in https://github.com/lummax/scalarelational/commit/4d76103ce0e7a0398edb07da82775095ee750da6.

- MySQL does not support the `MERGE INTO` syntax. As of https://github.com/lummax/scalarelational/commit/5a018f9cc44df533dc4fdc5a414551b1bd76b419 the fix is to set `MySQLDatastore.supportsMerge = false`.

- MySQL returns an `Array[Byte]` for the `BLOB` data type in violation of the JDBL 4.1 specification.

- MySQL has a maximum row size limit, so we can't use a `VARCHAR(65535)` multiple times or in combination with the `UNIQUE` index. The current "fix" is to introduce `ColumnLength` for all `String` fields in `AbstractTableSpec`. I dismissed the idea of a default maximum size of `X` for `MySQLDatastore` is it introduces "magic". In the long run it might be better to implement some kind of automatic length detection.

  The second problem is the usage of `BINARY` in `ObjectSerializationConverter`. The fix is to use a `BLOB` (again the MySQL `BLOB` issue).

- `IF NOT EXISTS` on `CREATE INDEX` seems to be only supported on H2. This restriction was removed.

- MySQL ignores `CASCADE` on `DROP TABLE`, so two of the cross reference test cases fail. I'm not certain what the best fix would be.

**Comment by darkfrog26** [comment]
@tindzk do you have any thoughts regarding `CASCADE` alternatives?

@lummax you might try something like this as the implementation if `CASCADE` is set:

`mysql> SET foreign_key_checks = 0;
mysql> drop table ...
mysql> SET foreign_key_checks = 1;`

**Comment by lummax** [comment]
@darkfrog26 Right, that seems to be the solution for MySQL. But as far as I understand it these settings would need to surround all of the `DROP` statements? Maybe I'm wrong.

Would you be ok with renaming the MySQL-Support to MariaDB-Support and stating a supported version in the readme?

I'll look into this and the other comments from @Kiwi and @tindzk this evening.

**Comment by darkfrog26** [comment]
@lummax you'd just override `def(drop: DropTable): List[CallableInstruction]` like this:

```scala
override def ddl(drop: DropColumn): List[CallableInstruction] = {
  List(CallableInstruction("SET foreign_key_checks = 0;"), super.ddl(drop).head, CallableInstruction("SET foreign_key_checks = 1;"))
}
```

**Comment by darkfrog26** [comment]
Yes, I'd be fine with renaming to MariaDB.

**Comment by darkfrog26** [comment]
Great job @lummax!
